students/alex_skrn/Lesson05/simple.py

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+
+"""Logging and debugging lesson."""
+
+import logging
+import datetime
+from logging import handlers
+
+
+format = "%(asctime)s %(filename)s:%(lineno)-4d %(levelname)s %(message)s"
+# Create a "formatter" using our format string
+formatter = logging.Formatter(format)
+
+# Setting up a formatter for the syslog logger - no time stamp
+format2 = "%(filename)s:%(lineno)-4d %(levelname)s %(message)s"
+syslog_formatter = logging.Formatter(format2)
+
+# Create a log message handler that sends output to the file 'todays_date.log'
+todays_date = str(datetime.date.today())
+file_handler = logging.FileHandler("{}.log".format(todays_date))
+file_handler.setLevel(logging.WARNING)
+# Set the formatter for this log message handler to the formatter we created above.
+file_handler.setFormatter(formatter)
+
+# default stream: sys.stderr stream one of two system streams that
+# get printed directly to the console
+console_handler = logging.StreamHandler()
+console_handler.setLevel(logging.DEBUG)
+console_handler.setFormatter(formatter)
+
+# Setting up a syslog server logger for Windows
+syslog_handler_w = logging.handlers.DatagramHandler("127.0.0.1", 514)
+syslog_handler_w.setLevel(logging.ERROR)
+syslog_handler_w.setFormatter(syslog_formatter)
+
+# # Setting up a syslog server logger for Mac
+# syslog_handler_m = logging.SysLogHandler()
+# syslog_handler_m.setLevel(logging.ERROR)
+# syslog_handler_m.setFormatter(syslog_formatter)
+
+# Get the "root" logger.
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+# Add our file_handler to the "root" logger's handlers.
+logger.addHandler(file_handler)
+logger.addHandler(console_handler)
+logger.addHandler(syslog_handler_w)
+# logger.addHandler(syslog_handler_m)
+
+
+def my_fun(n):
+    for i in range(0, n):
+        logging.debug(i)
+        if i == 50:
+            logging.warning("The value of i is 50.")
+        try:
+            i / (50 - i)
+        except ZeroDivisionError:
+            msg = "Tried to divide by zero. Var i was {}. Recovered gracefully"
+            logging.error(msg.format(i))
+
```

Comment:
For some more fun, See if you can add a bit more functionality to create more logging and add multiple threads.  Then Provide a uuuid in your logging that is specific to a single run of the thread.  sudo code way of explaining it:
```
def start_thread_to_execute_function_with_uuid_for_the_log():
    # Three logging levels done like you have above but a uuid included that is common to all of them

for i in range(0, n):
   start_thread_to_execute_function_with_uuid_for_the_log()
   # Make sure the threads exits after its done with execution
```
You should now be able to grep your file output for a uuid and see the three logs you created per execution.

Your code is always really well done so I figured I'd add a more practical challenge with this.

Reply:
If you'd rather skip it that is of course fine as well :)

Reply:
Hi Michael, I created a PR for this with my initial steps and some questions. More questions actually than solutions.
https://github.com/UWPCE-PythonCert-ClassRepos/Sp2018-Online/pull/104